File: DataStructure/DisjointSet/ExtDisjointSet.cpp
#include <vector>

using namespace std;

struct ExtDisjointSet {
  vector<int> p, s, t;

  static inline int RE(int i) { return ~i; }  // !!

  void init(int n) {
    p.resize(n);
    s.resize(n);
    t.resize(n);
    for (int i = 0; i < n; ++i) {
      p[i] = i;
      s[i] = 1;
      t[i] = 0;
    }
  }

  int getp(int i) {
    if (i < 0) {
      return RE(getp(RE(i)));
    } else if (i == p[i]) {
      return i;
    } else {
      return p[i] = getp(p[i]);
    }
  }

  int setp(int i, int j) {
    i = getp(i);
    j = getp(j);
    if (i == j) {
      return 0;
    } else if (i == RE(j)) {
      return -1;
    } else {
      if (i < 0) {
        i = RE(i);
        j = RE(j);
      }
      p[i] = j;
      if (j >= 0) {
        s[j] += s[i];
        t[j] += t[i];
      } else {
        s[RE(j)] += t[i];
        t[RE(j)] += s[i];
      }
      return 1;
    }
  }
} eds;

File: DataStructure/DisjointSet/ExtDisjointSet.rs

pub struct ExtDisjointSet(Vec<i32>);

impl ExtDisjointSet {
    pub fn new(n: i32) -> Self {
        Self((0..n).collect())
    }

    pub fn getp(&mut self, x: i32) -> i32 {
        if x < 0 {
            !self.getp(!x)
        } else {
            let i = x as usize;
            if self.0[i] == x {
                x
            } else {
                self.0[i] = self.getp(self.0[i]);
                self.0[i]
            }
        }
    }

    pub fn setp(&mut self, x: i32, y: i32) -> Result<(i32, i32), bool> {
        let x = self.getp(x);
        let y = self.getp(y);
        if x == y {
            Err(true)
        } else if x == !y {
            Err(false)
        } else {
            let (x, y) = if x >= 0 { (x, y) } else { (!x, !y) };
            self.0[x as usize] = y;
            Ok((x, y))
        }
    }
}
